docs/prefect integration (#3037)
dlt-hub/dlt: docs/website/docs/plus/production/prefect-integration.md

---
title: "Prefect Integration"
description: Run pipelines with Prefect
keywords: ["prefect", "retry", "logs", "artifacts", "progress", "callbacks", "source decomposition", "parallelization"]
---
# Prefect Integration

dlt+ offers a few tools and helpers to make running dlt pipelines in prefect a seamless experience.

## Key features

> - **Prefect Collector:** a dedicated way to do real-time [progress monitoring] and summary reports after each pipeline stage

Reviewer: why `[progress monitoring]` is in braces?

Author: thx, i thought maybe this could be a link but removed it ultimately: https://dlthub.com/docs/general-usage/pipeline#monitor-the-loading-progress

- **Source decomposition flows:** Create tasks and flows that will execute your pipelines in parallel with the right settings

## Prefect Collector

The `PrefectCollector` creates rich artifacts in the Prefect UI to monitor dlt pipeline progress and results. Simply pass it to your pipeline's `progress` parameter to enable detailed monitoring.

```py
from dlt_plus._runner.prefect_collector import PrefectCollector

@task
def my_dlt_task():
    pipeline = dlt.pipeline(
        pipeline_name="my_pipeline",
        destination="duckdb",
        dataset_name="my_dataset",
        progress=PrefectCollector()
    )
    
    load_info = pipeline.run(source)
```

### Progress and Summary Reports

If the task is executed the `PrefectCollector` will create progress artifacts that will update
during a stage as well as summary report after a pipeline step has completed.
For example during extract you will see how many resources have been extracted and once the extract 
is completed you will see how many rows have been processed for each resource: 

![Prefect Extract Artifacts](https://storage.googleapis.com/dlt-blog-images/docs-prefect-extract-artifact.png)

> Also, each summary report contains basic information about the pipeline, the destination the execution 

Reviewer: ```suggestion
Each summary report includes basic information about the pipeline, the destination, the execution, and system resource usage (CPU and memory).
```

> environment and system resources (CPU, memory) usage.

Reviewer: ```suggestion
```

### Schema Change Reports

> The `PrefectCollector` will also create artifacts when schema changes are detected.

Reviewer: should we elaborate on the artifacts? what's inside?

This is useful to see which tables and columns have changed between runs.

![Prefect Schema Change Detection](https://storage.googleapis.com/dlt-blog-images/docs-prefect-schema-change-artifact.png)

### Logging

Prefect has built-in functionality to [include logs from other libraries](https://docs.prefect.io/v3/advanced/logging-customization#include-logs-from-other-libraries) and display them as part of their UI.

You can tell prefect to include `dlt`'s logs by setting the corresponding prefect environment variable, for example by adding this to your `.env` file:

> ```sh

Reviewer: nit: Could you add the `secrets.toml` version as well?

PREFECT_LOGGING_EXTRA_LOGGERS=dlt
```

You should now see dlt's logs in the prefect UI and be able to [query them with the prefect CLI](https://docs.prefect.io/v3/how-to-guides/workflows/add-logging#access-logs-from-the-command-line).

## Runner integration

> The `PrefectCollector` integrates seamlessly with the [dlt+ runner](../production/pipeline-runner.md). 

Reviewer: ```suggestion
The `PrefectCollector` integrates seamlessly with the [dlt+ Runner](../production/pipeline-runner.md). 
```

If you are running your pipelines with the dlt+ runner, information about the pipelines run configuration and the trace pipeline 
will also be included in the artifacts in the `Run Configuration` section.

### Pipeline Retries

> Prefect retry-mechanism is not a perfect fit for dlt pipelines.

Reviewer: I suggest first to mention that dlt+ Runner fixes the problem and then explain it :) 

```
The dlt+ Runner provides a retry configuration that ensures pipeline state and intermediate results are preserved across retry attempts.

This is important because Prefect’s default retry mechanism is not optimized for dlt pipelines. During execution, dlt generates intermediate files in the pipeline’s working directory. If a run fails and Prefect retries it, those files may not be available anymore. For example, if the retry happens on a different worker node or inside an ephemeral Docker container.

To do so, ...
```

    )
    source = sql_database(schema="my_schema")

    # create a subflow that will run each resource in a dedicated pipeline
    run_pipeline_with_parallel_tables = create_run_pipeline_flow(
        pipeline=pipeline,
        source=source,
        decompose=True,
        flow_name="my_parallelized_flow",
    )

    # run it, passing args that you would usually pass to the pipeline.run() method
    run_pipeline_with_parallel_tables(write_disposition="append")

if __name__ == "__main__":
    my_main_flow()

```

This will derive a task-pipeline with a deterministic name for each resource in the source and run them in parallel.
Because they have different names, pipelines can run in parallel without interfering with each other.

You can read more about source decomposition and what to watch out for when using it on the [performance page](../../reference/performance#source-decomposition-for-serial-and-parallel-resource-execution).

![Prefect Source Decomposition Flow](https://storage.googleapis.com/dlt-blog-images/docs-prefect-source-decomposition.png)

:::note
For stability reasons, this actually runs one resource alone and then all others in parallel.

> This is because otherwise, on the first run, all resources would try to create the same dlt-tables.

Reviewer: ```suggestion
This prevents multiple resources from attempting to create the same dlt tables simultaneously during the initial run.
```

:::

> :::warning

Author: @ShreyasGS read this

The flow created by this helper must be executed with the thread pool task runner.
Other runners, which try to distribute tasks across multiple machines won't work with these helpers, because they
pass pipeline and source objects as arguments, which can not be pickled or serialized in order to be sent to other machines.
If want to build tasks that run on multiple machines, you will have to initialize both pipeline and sources inside those tasks.
:::

## On parallelization

> Both `dlt` and prefect can be configured to use multithreading and multiprocessing, but at different stages of the pipeline. 

Reviewer: ```suggestion
Both `dlt` and Prefect can be configured to use multithreading and multiprocessing, but at different stages of the pipeline. 
```
